localstack/services/cloudformation/engine/template_deployer.py

```diff
@@ -397,19 +389,21 @@ def resolve_refs_recursively(stack_name, resources, value):
 
 
 @prevent_stack_overflow(match_parameters=True)
-def _resolve_refs_recursively(stack_name, resources, value: dict | list | str | bytes | None):
+def _resolve_refs_recursively(
+    stack_name: str, resources: dict, mappings: dict, value: dict | list | str | bytes | None
+):
     if isinstance(value, dict):
         keys_list = list(value.keys())
         stripped_fn_lower = keys_list[0].lower().split("::")[-1] if len(keys_list) == 1 else None
 
         # process special operators
         if keys_list == ["Ref"]:
-            ref = resolve_ref(stack_name, resources, value["Ref"], attribute="Ref")
+            ref = resolve_ref(stack_name, resources, mappings, value["Ref"], attribute="Ref")
             if ref is None:
                 msg = 'Unable to resolve Ref for resource "%s" (yet)' % value["Ref"]
                 LOG.debug("%s - %s", msg, resources.get(value["Ref"]) or set(resources.keys()))
                 raise DependencyNotYetSatisfied(resource_ids=value["Ref"], message=msg)
-            ref = resolve_refs_recursively(stack_name, resources, ref)
+            ref = resolve_refs_recursively(stack_name, resources, mappings, ref)
             return ref
 
         if stripped_fn_lower == "getatt":
```

Comment:
Yup, that will be a whole intrinsic function rework. Unfortunately we're not quite there yet were we can replace the intrinsic function resolving. The idea we had here was to build a sort of "hydrated" stack that contains the whole dependency graph and a somwhat "rich" API to navigate it & resolve refs, etc.

Reply:
The current approach is also a bit of a problem because we allow things that we shouldn't. A lot of sections / Intrinsic functions only allow a subset of other intrinsic functions to be contained in their fields, which we currently   can't easily model with this construct.